src/Compilers/Core/Portable/PEWriter/PdbWriter.cs

```diff
@@ -54,15 +55,59 @@ internal sealed class PdbWriter : IDisposable
         private uint[] _sequencePointEndLines;
         private uint[] _sequencePointEndColumns;
 
-        public PdbWriter(Stream stream, string fileName, Func<object> symWriterFactory = null)
+        [DllImport("shlwapi.dll", ExactSpelling = true, CharSet = CharSet.Unicode)]
+        private extern static IStream SHCreateMemStream([In] IntPtr pInit, [In] uint cbInit);
+
+        public PdbWriter(string fileName, Func<object> symWriterFactory = null)
         {
-            Debug.Assert(stream != null);
-            _stream = stream;
             _fileName = fileName;
             _symWriterFactory = symWriterFactory;
             CreateSequencePointBuffers(capacity: 64);
         }
 
+        public void WriteTo(Stream stream)
+        {
+            Debug.Assert(_nativeStream != null);
+            Debug.Assert(_symWriter != null);
+
+            try
+            {
+                // SymWriter flushes data to the native stream on close:
+                _symWriter.Close();
+                _symWriter = null;
+
+                var statStg = default(STATSTG);
+                _nativeStream.Stat(out statStg, grfStatFlag: 3 /*STATFLAG_NONAME | STATFLAG_NOOPEN*/);
+                _nativeStream.Seek(0L, 0, IntPtr.Zero);
+
+                int size = (int)statStg.cbSize;
+
+                // If the target stream allows seeking set its length upfront.
+                // When writing to a large file, it helps to give a hint to the OS how big the file is going to be.
+                if (stream.CanSeek)
+                {
+                    stream.SetLength(stream.Position + size);
+                }
+
+                // Copy unmanagedStream contents to stream in chunks
+                const int maxChunkSize = 4096;
+                var buffer = new byte[Math.Min(size, maxChunkSize)];
+                while (size > 0)
+                {
+                    int chunkSize = Math.Min(size, buffer.Length);
+                    _nativeStream.Read(buffer, chunkSize, IntPtr.Zero);
```

Comment:
Good point. We query the IStream for size, so it should be ok. However, you're right that the common pattern is to check the bytes actually read. I'm gonna change it to do so.